src/coreclr/src/gc/gc.cpp

```diff
@@ -2060,6 +2068,42 @@ uint8_t* tree_search (uint8_t* tree, uint8_t* old_address);
 
 #ifdef USE_INTROSORT
 #define _sort introsort::sort
+#elif defined(USE_VXSORT)
+#define _sort vxsort
+namespace std
+{
+    template <class _Ty>
+    class numeric_limits
+    {
+    public:
+        static _Ty Max()
+        {
+            return _Ty();
+        }
+    };
+
+    template <>
+    class numeric_limits<int64_t>
+    {
+    public:
+        static int64_t Max()
+        {
+            return LLONG_MAX;
+        }
+    };
+}
+void vxsort(uint8_t** low, uint8_t** high, unsigned int depth)
+{
```

Comment:
Do we need "is avx available" check here and fallback to the non-vectorized sort ?

Reply:
Yeap, unless you'd like me/Peter to copy the rest of the non-vectorized partitioning back into the same sort.

It already shares the heap-sort + insertion-sort code from the original as a fallback just in case we run into one of the existing edge-cases, as with standard introsort.

This should reduce duplication, as I literally copied 2/3 of introsort into vxsort from the get go.

Reply:
With clang/gcc I've previously used [builtins](https://gcc.gnu.org/onlinedocs/gcc/x86-Built-in-Functions.html):
```c 
if (__builtin_cpu_supports("avx2")) {
  // AVX2
} else {
  // SAD PANDA
}
```

To cleanly detect this.

Pretty sure you'll have to do something else for MSVC.

Reply:
Do we have prior art for this? I found EEJitManager::SetCpuInfo which detects the AVX2 instruction set for the JIT.

Reply:
For Unix, `__builtin_cpu_supports("avx2")` looks pretty nice. For Windows, I guess it will need to call `GetEnabledXStateFeatures` and check for `XSTATE_MASK_AVX`

Reply:
The general mechanism I'm familiar with is using function multi-versioning, this is supported in gcc since forever, and in clang from 7.0 onwards if I'm not mistaken.

The basic idea would be that a function can be decorated with:
```c
__attribute__ ((target ("default")))
int foo () {

  return 1;
}
__attribute__ ((target ("avx2")))
int foo () {
  return 2;
}
```

And you get an indirect jump embedded in the code stream courtesy of the compiler.

I'm pretty sure this does not exist for MSVC, where you need to hand-code such things, but then again, you have more apt MSVC gurus around you, so they might know something I don't. But it seems that if you would like to avoid depending on the EEJitManager, I could add support for this myself for clang/gcc the normal mechanisms, and for MSVC by following what seems to be the [standard practice](https://docs.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?redirectedfrom=MSDN&view=vs-2019) inside vxsort, that way I can switch from AVX2, AVX512 when my AVX512 support is complete.

Would you like me to take care of all of this internally and expose one sort functionality that does everything?

Reply:
Is there any reason we can't use the cached checks the VM already has from the HWIntrinsics feature?

Reply:
We want to support building GC into separate .dll (aka LocalGC).

Reply:
I'll also add that you don't detect AVX512 yet either, which I'm working on in a vxsort branch, regardless if @PeterSolMS ends up taking it.